devmode breaks when updating pom.xml

### Describe the bug

in past I could add a maven dependency and quakrus devmode would restart.

now when I do it devmode crashes.

```
ERROR: transport error 202: bind failed: Address already in use
ERROR: JDWP Transport dt_socket failed to initialize, TRANSPORT_INIT(510)
JDWP exit error AGENT_ERROR_TRANSPORT_INIT(197): No transports initialized [src/jdk.jdwp.agent/share/native/libjdwp/debugInit.c:700]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:57 min
[INFO] Finished at: 2024-05-26T23:38:51+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:3.10.2:dev (default-cli) on project jconqbot: Failed to run: Dev mode process did not complete successfully -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```


### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/core/deployment/src/main/java/io/quarkus/deployment/dev/QuarkusDevModeLauncher.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/dev/QuarkusDevModeLauncher.java
@@ -86,12 +86,6 @@ public abstract class QuarkusDevModeLauncher {
         }
 
         @SuppressWarnings("unchecked")
-        public B debugPortOk(Boolean debugPortOk) {
-            QuarkusDevModeLauncher.this.debugPortOk = debugPortOk;
-            return (B) this;
-        }
-
-        @SuppressWarnings("unchecked")
         public B suspend(String suspend) {
             QuarkusDevModeLauncher.this.suspend = suspend;
             return (B) this;
@@ -303,10 +297,10 @@ public abstract class QuarkusDevModeLauncher {
 
     private List<String> args = new ArrayList<>(0);
     private String debug;
-    private Boolean debugPortOk;
     private String suspend;
     private String debugHost = "localhost";
     private String debugPort = "5005";
+    private String actualDebugPort;
     private File projectDir;
     private File buildDir;
     private File outputDir;
@@ -390,12 +384,13 @@ public abstract class QuarkusDevModeLauncher {
 
         if (debug != null && debug.equalsIgnoreCase("client")) {
             args.add("-agentlib:jdwp=transport=dt_socket,address=" + debugHost + ":" + port + ",server=n,suspend=" + suspend);
+            actualDebugPort = String.valueOf(port);
         } else if (debug == null || !debug.equalsIgnoreCase("false")) {
             // if the debug port is used, we want to make an effort to pick another one
             // if we can't find an open port, we don't fail the process launch, we just don't enable debugging
             // Furthermore, we don't check this on restarts, as the previous process is still running
             boolean warnAboutChange = false;
-            if (debugPortOk == null) {
+            if (actualDebugPort == null) {
                 int tries = 0;
                 while (true) {
                     boolean isPortUsed;
@@ -408,20 +403,19 @@ public abstract class QuarkusDevModeLauncher {
                         isPortUsed = false;
                     }
                     if (!isPortUsed) {
-                        debugPortOk = true;
+                        actualDebugPort = String.valueOf(port);
                         break;
                     }
                     if (++tries >= 5) {
-                        debugPortOk = false;
                         break;
                     } else {
                         port = getRandomPort();
                     }
                 }
             }
-            if (debugPortOk) {
+            if (actualDebugPort != null) {
                 if (warnAboutChange) {
-                    warn("Changed debug port to " + port + " because of a port conflict");
+                    warn("Changed debug port to " + actualDebugPort + " because of a port conflict");
                 }
                 args.add("-agentlib:jdwp=transport=dt_socket,address=" + debugHost + ":" + port + ",server=y,suspend="
                         + suspend);
@@ -547,8 +541,8 @@ public abstract class QuarkusDevModeLauncher {
         return args;
     }
 
-    public Boolean getDebugPortOk() {
-        return debugPortOk;
+    public String getActualDebugPort() {
+        return actualDebugPort;
     }
 
     protected abstract boolean isDebugEnabled();

--- a/devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java
+++ b/devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java
@@ -478,15 +478,19 @@ public class DevMojo extends AbstractMojo {
                     }
                     if (!changed.isEmpty()) {
                         getLog().info("Changes detected to " + changed + ", restarting dev mode");
+
+                        // stop the runner before we build the new one as the debug port being free
+                        // is tested when building the runner
+                        runner.stop();
+
                         final DevModeRunner newRunner;
                         try {
                             bootstrapId = handleAutoCompile();
-                            newRunner = new DevModeRunner(runner.launcher.getDebugPortOk(), bootstrapId);
+                            newRunner = new DevModeRunner(runner.launcher.getActualDebugPort(), bootstrapId);
                         } catch (Exception e) {
                             getLog().info("Could not load changed pom.xml file, changes not applied", e);
                             continue;
                         }
-                        runner.stop();
                         newRunner.run();
                         runner = newRunner;
                     }
@@ -1171,8 +1175,8 @@ public class DevMojo extends AbstractMojo {
             launcher = newLauncher(null, bootstrapId);
         }
 
-        private DevModeRunner(Boolean debugPortOk, String bootstrapId) throws Exception {
-            launcher = newLauncher(debugPortOk, bootstrapId);
+        private DevModeRunner(String actualDebugPort, String bootstrapId) throws Exception {
+            launcher = newLauncher(actualDebugPort, bootstrapId);
         }
 
         Collection<Path> pomFiles() {
@@ -1226,7 +1230,7 @@ public class DevMojo extends AbstractMojo {
         }
     }
 
-    private QuarkusDevModeLauncher newLauncher(Boolean debugPortOk, String bootstrapId) throws Exception {
+    private QuarkusDevModeLauncher newLauncher(String actualDebugPort, String bootstrapId) throws Exception {
         String java = null;
         // See if a toolchain is configured
         if (toolchainManager != null) {
@@ -1244,8 +1248,7 @@ public class DevMojo extends AbstractMojo {
                 .suspend(suspend)
                 .debug(debug)
                 .debugHost(debugHost)
-                .debugPort(debugPort)
-                .debugPortOk(debugPortOk)
+                .debugPort(actualDebugPort)
                 .deleteDevJar(deleteDevJar);
 
         setJvmArgs(builder);

